### ISSUE TITLE
Can I extract attribute values using OpenScraping?

### ISSUE THREAD

**Issue by Porkstone** [microsoft/openscraping-lib-csharp#6]
`<a href="test">dfsdfsdf</a>`

I have tried  systax like this //a/@href 
It just returns the contents of the anchor tag, but I'm looking for "test" in the href attribute.

Is this possible?

**Comment by zmarty** [comment]
It should be possible. I will give it a try.

**Comment by zmarty** [comment]
It seems to work just fine:

```csharp
var configJson = "{ 'link': '//a/@href' }";
var html = "<html><body><a href='test'>dfsdfsdf</a></body></html>";

var config = StructuredDataConfig.ParseJsonString(configJson);
var extractor = new StructuredDataExtractor(config);
var result = extractor.Extract(html);
var json = JsonConvert.SerializeObject(result, Formatting.Indented);
```

The json output is:

```javascript
{
  "link": "dfsdfsdf"
}
```

You can even force an array like this:
```cspharp
var configJson = "{ 'links': {'_xpath': '//a/@href', '_forceArray': true} }";
var html = "<html><body><a href='test'>dfsdfsdf</a></body></html>";

var config = StructuredDataConfig.ParseJsonString(configJson);
var extractor = new StructuredDataExtractor(config);
var result = extractor.Extract(html);
var json = JsonConvert.SerializeObject(result, Formatting.Indented);
```

```javascript
{
  "links": [
    "dfsdfsdf"
  ]
}
```

**Comment by zmarty** [comment]
Sorry, I mistakenly closed this item. This is still a problem.

**Comment by zmarty** [comment]
OpenScraping depends on Html Agility Pack, which does not support attribute selection, as explained [here](http://htmlagilitypack.codeplex.com/discussions/1720).

What I can do is modify the code to support simple attribute selection, so your particular example would work. Let me know if you still need this and I will make the change.

**Comment by nickhaughton** [comment]
I could use this, I'm happy to do the work if needs be. @zmarty

**Comment by mcskelle** [comment]
I'm pretty happy using the library but really really need to extract links using /a/@href or some other method in the lib. Any chance this will be available soon?

**Comment by cbracht** [comment]
@zmarty The library is working great so far but I'm in need to extract the href attribute of a link as well. A workaround would be great.

**Comment by zmarty** [comment]
@cbracht Acknowledged, will work on it.

**Comment by cbracht** [comment]
Thank you, your work is appreciated!

**Comment by marcel-silva** [comment]
I made it work like this:

    public class LinkAttributeValue : ITransformationFromHtml
    {
        public object Transform(Dictionary<string, object> settings, HtmlAgilityPack.HtmlNode node, List<HtmlAgilityPack.HtmlNode> logicalParents)
        {
            if (node != null)
            {
                string attrName = "href";
                string fallBack = "#";
                if (settings != null && settings["_attrName"] != null && ((JValue)settings["_attrName"]).Type == JTokenType.String)
                {
                    attrName = ((JValue)settings["_attrName"]).ToObject<string>();
                }
                if (settings != null && settings["_fallBack"] != null && ((JValue)settings["_fallBack"]).Type == JTokenType.String)
                {
                    fallBack = ((JValue)settings["_fallBack"]).ToObject<string>();
                }

                var href = node.GetAttributeValue(attrName, fallBack);
                return href;
            }

            return null;
        }
    }



And to use:

          var html = "<html><body><h1>Article title</h1><div class='article'><a href='link1.com'>link1</a></div><div class='article'><a href='link2.com'>link2</a></div></body></html>";

            var configJson = @"{
                'teams': {
                    '_xpath': '//div[contains(@class, \'article\')]',
                    'name': './/a',
                    'link': {
                      '_xpath': './/a',
                      '_transformations': [
                        {
                            '_type': 'LinkAttributeValue',
                          '_attrName': 'href',
                          '_fallBack': '#'
                        }
                      ]
                    }
                }
            }";`

**Comment by zmarty** [comment]
Thanks for the workaround! I am working on a permanent solution in another branch. Will merge to master as soon as it's ready.

**Comment by avi22228** [comment]
@zmarty Do you have a fix for this ?

Thanks for the good work.

**Comment by zmarty** [comment]
Not yet, sorry. I did not have time to finish it.

**Comment by avi22228** [comment]
No worries. Thanks for the prompt reply. Hack posted by @marcel-silva is working fine for now.

When you release the next version, please also update the dependencies, currently it is interfering with other stuff.

Thank you..

**Comment by pldmgg** [comment]
Just wanted to chime in that I just ran into this as well. Using @marcel-silva workaround for time being. Thanks again for making this! It's very handy.

**Comment by shawnshaddock** [comment]
I also really need this fix. Is it still being worked on?

**Comment by shawnshaddock** [comment]
I have fixed this issue in pull request #14

**Comment by zmarty** [comment]
Thank you @shawnshaddock, this is now in master.

**Comment by zmarty** [comment]
@shawnshaddock Reopening since checkin #14 breaks transformations such as CastToIntegerTransformation. I am working on a fix, since I need this for a project.

**Comment by zmarty** [comment]
Fixed (hopefully permanently) through #21
